pydatastructs/trees/heaps.py

```diff
@@ -398,3 +398,179 @@ def delete(self, node):
+            self.min_node = n
+        self.total_nodes += 1
+        return n
+
+    # modify the key of some node in the heap in O(1) time
+    def decrease_key(self, x, k):
+        if k > x.key:
+            return None
+        x.key = k
+        y = x.parent
+        if y is not None and x.key < y.key:
+            self.cut(x, y)
+            self.cascading_cut(y)
+        if x.key < self.min_node.key:
+            self.min_node = x
+
+    # merge two fibonacci heaps in O(1) time by concatenating the root lists
+    # the root of the new root list becomes equal to the first list and the second
+    # list is simply appended to the end (then the proper min node is determined)
+    def merge(self, h2):
+        H = FibonacciHeap()
+        H.root_list, H.min_node = self.root_list, self.min_node
+        # fix pointers when merging the two heaps
+        last = h2.root_list.left
+        h2.root_list.left = H.root_list.left
+        H.root_list.left.right = h2.root_list
+        H.root_list.left = last
+        H.root_list.left.right = H.root_list
+        # update min node if needed
+        if h2.min_node.key < H.min_node.key:
+            H.min_node = h2.min_node
+        # update total nodes
+        H.total_nodes = self.total_nodes + h2.total_nodes
+        return H
+
+    # if a child node becomes smaller than its parent node we
+    # cut this child node off and bring it up to the root list
+    def cut(self, x, y):
+        self.remove_from_child_list(y, x)
+        y.degree -= 1
+        self.merge_with_root_list(x)
+        x.parent = None
+        x.mark = False
+
+    # cascading cut of parent node to obtain good time bounds
+    def cascading_cut(self, y):
+        z = y.parent
+        if z is not None:
+            if y.mark is False:
+                y.mark = True
+            else:
+                self.cut(y, z)
+                self.cascading_cut(z)
+
+    # combine root nodes of equal degree to consolidate the heap
+    # by creating a list of unordered binomial trees
+    def consolidate(self):
+        A = [None] * int(math.log(self.total_nodes) * 2)
+        nodes = [w for w in self.iterate(self.root_list)]
+        for w in range(0, len(nodes)):
+            x = nodes[w]
+            d = x.degree
+            while A[d] != None:
+                y = A[d]
+                if x.key > y.key:
+                    temp = x
+                    x, y = y, temp
+                self.heap_link(y, x)
+                A[d] = None
+                d += 1
+            A[d] = x
+        # find new min node - no need to reconstruct new root list below
+        # because root list was iteratively changing as we were moving
+        # nodes around in the above loop
+        for i in range(0, len(A)):
+            if A[i] is not None:
+                if A[i].key < self.min_node.key:
+                    self.min_node = A[i]
+
+    # actual linking of one node to another in the root list
+    # while also updating the child linked list

```

Comment:
Instead of comments please add doc strings. See the existing code.

Reply:
I have added doc strings